yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala

```diff
@@ -124,27 +124,32 @@ private[spark] class YarnClientSchedulerBackend(
    * If the application has exited for any reason, stop the SparkContext.
    * This assumes both `client` and `appId` have already been set.
    */
-  private def asyncMonitorApplication(): Thread = {
+  private def asyncMonitorApplication(): MonitorApplicationThread = {
     assert(client != null && appId != null, "Application has not been submitted yet!")
-    val t = new Thread {
-      override def run() {
-        val (state, _) = client.monitorApplication(appId, logApplicationReport = false)
-        logError(s"Yarn application has already exited with state $state!")
-        sc.stop()
-        Thread.currentThread().interrupt()
-      }
-    }
+    val t = new MonitorApplicationThread()
     t.setName("Yarn application state monitor")
     t.setDaemon(true)
     t
   }
 
+  private class  MonitorApplicationThread extends Thread{
+    override def run() {
+      val (state, _) = client.monitorApplication(appId, logApplicationReport = false)
+      logError(s"Yarn application has already exited with state $state!")
+      sc.stop()
+      Thread.currentThread().interrupt()
+    }
+    def stopMonitor() {
+      client.stopMonitorApplication()
+    }
+  }
+
   /**
    * Stop the scheduler. This assumes `start()` has already been called.
    */
   override def stop() {
     assert(client != null, "Attempted to stop this scheduler before starting it!")
-    monitorThread.interrupt()
+    monitorThread.stopMonitor()
```

Comment:
The narrowest fix is just to catch `InterruptedException` here and ignore it and/or log it.